piet-common/src/cg_back.rs

```diff
@@ -0,0 +1,174 @@
+// allows e.g. raw_data[dst_off + x * 4 + 2] = buf[src_off + x * 4 + 0];
+#![allow(clippy::identity_op)]
+
+//! Support for piet CoreGraphics back-end.
+
+use std::marker::PhantomData;
+use std::path::Path;
+#[cfg(feature = "png")]
+use std::{fs::File, io::BufWriter};
+
+use core_graphics::{color_space::CGColorSpace, context::CGContext, image::CGImage};
+#[cfg(feature = "png")]
+use png::{ColorType, Encoder};
+
+use piet::{ErrorKind, ImageFormat};
+#[doc(hidden)]
+pub use piet_coregraphics::*;
+
+/// The `RenderContext` for the CoreGraphics backend, which is selected.
+pub type Piet<'a> = CoreGraphicsContext<'a>;
+
+/// The associated brush type for this backend.
+///
+/// This type matches `RenderContext::Brush`
+pub type Brush = piet_coregraphics::Brush;
+
+/// The associated text factory for this backend.
+///
+/// This type matches `RenderContext::Text`
+pub type PietText<'a> = CoreGraphicsText<'a>;
+
+/// The associated font type for this backend.
+///
+/// This type matches `RenderContext::Text::Font`
+pub type PietFont = CoreGraphicsFont;
+
+/// The associated font builder for this backend.
+///
+/// This type matches `RenderContext::Text::FontBuilder`
+pub type PietFontBuilder<'a> = CoreGraphicsFontBuilder;
+
+/// The associated text layout type for this backend.
+///
+/// This type matches `RenderContext::Text::TextLayout`
+pub type PietTextLayout = CoreGraphicsTextLayout;
+
+/// The associated text layout builder for this backend.
+///
+/// This type matches `RenderContext::Text::TextLayoutBuilder`
+pub type PietTextLayoutBuilder<'a> = CoreGraphicsTextLayout;
+
+/// The associated image type for this backend.
+///
+/// This type matches `RenderContext::Image`
+pub type Image = CGImage;
+
+/// A struct that can be used to create bitmap render contexts.
+pub struct Device;
+
+/// A struct provides a `RenderContext` and then can have its bitmap extracted.
+pub struct BitmapTarget<'a> {
+    ctx: CGContext,
+    height: f64,
+    phantom: PhantomData<&'a ()>,
+}
+
+impl Device {
+    /// Create a new device.
+    pub fn new() -> Result<Device, piet::Error> {
+        Ok(Device)
+    }
+
+    /// Create a new bitmap target.
+    pub fn bitmap_target(
+        &mut self,
+        width: usize,
+        height: usize,
+        pix_scale: f64,
+    ) -> Result<BitmapTarget, piet::Error> {
+        let ctx = CGContext::create_bitmap_context(
+            None,
+            width,
+            height,
+            8,
+            0,
+            &CGColorSpace::create_device_rgb(),
+            core_graphics::base::kCGImageAlphaPremultipliedLast,
+        );
+        ctx.scale(pix_scale, pix_scale);
+        let height = height as f64 * pix_scale;
```

Comment:
Doesn't `bitmap_target` take the height in pixels? Looking at other platforms it seems to. Then if you want the logical height you gotta divide.

Reply:
I think it's the height in device points? Isn't that why we pass in the pix_scale argument?